Implement Arm64 Unwind Codes for Windows

scott-ferguson-unity opened the pull request:
The unwind data is registered using the same functions as for
Windows Amd64, however the actual unwind data is completely
different.

Windows Arm64 records the unwind data as a 1-to-1 mapping between
the codes and the instructions in the prolog and epilog.  This is
to support unwinding in the middle of the epilog or prolog.  It's not
clear if we need this support in mono, but it has been faithfully
implemented.  This includes emitting nop unwind codes for instructions
in the header that aren't stack related.

The unwind codes also support different formats for reduce the amount
of data stored and optimize for the case were the prolog and epilog
are mirror images.  This was not the case previously, mono would restore
the preserved registers in the same order, instead of the reverse.  So
a new function was added to emit them in the reverse order.

Another important case is last managed frame methods (LMF).  These are
managed methods that will transition into native code.  These methods
always spill all of the preserved registers to stack in case they
need to recover from a managed exception, but only restore the
registers they use.  So these functions can never use the optimized
forms.

- Should this pull request have release notes?
  - [x] Yes
  - [ ] No
- Do these changes need to be back ported?
  - [x] Yes
  - [ ] No
- Do these changes need to be upstreamed to [mono/mono](https://github.com/mono/mono) or [dotnet/runtime](https://github.com/dotnet/runtime) repositories?
  - [ ] Yes
  - [x] No

Reviewers: please consider these questions as well! :heart:

**Release notes**

Fixed UUM-49668 @scott-ferguson-unity 
Mono: Add support for stack unwind code registration on Windows Arm64

<!-- Most pull requests should have release notes.

Use Internal for release notes that should not be public.

Other options: Changed, Improved, Feature.
-->

**Backports**
* 2023.3

**Upstream**
Upstream has no support for Windows Arm64 - we would need to upstream all of our  Windows Arm64 changes.

<!-- Use this section if the pull request requires other changes in the Unity repository.
**Unity repository changes**

List any Unity repository PRs.
-->

scott-ferguson-unity committed:
Implement unwind code registration for Windows Arm64

The unwind data is registered using the same functions as for
Windows Amd64, however the actual unwind data is completely
different.

Windows Arm64 records the unwind data as a 1-to-1 mapping between
the codes and the instructions in the prolog and epilog.  This is
to support unwinding in the middle of the epilog or prolog.  It's not
clear if we need this support in mono, but it has been faithfully
implemented.  This includes emitting nop unwind codes for instructions
in the header that aren't stack related.

The unwind codes also support different formats for reduce the amount
of data stored and optimize for the case were the prolog and epilog
are mirror images.  This was not the case previously, mono would restore
the preserved registers in the same order, instead of the reverse.  So
a new function was added to emit them in the reverse order.

Another important case is last managed frame methods (LMF).  These are
managed methods that will transition into native code.  These methods
always spill all of the preserved registers to stack in case they
need to recover from a managed exception, but only restore the
registers they use.  So these functions can never use the optimized
forms.

joncham commented on mono/mini/mini-arm64.c:
```diff
@@ -5114,6 +5114,66 @@ emit_store_regset_cfa (MonoCompile *cfg, guint8 *code, guint64 regs, int basereg
+		}
+	}
+
+	for (i = 32; i >= 0; --i) {
```
Should this be `i = 31` reverse the loop above?

scott-ferguson-unity replied:
Yes it should be 31 - fixed in https://github.com/Unity-Technologies/mono/pull/1960/commits/9b8e291ee6dde9c51871bf05de6ad195e1d48f01

joncham commented on mono/mini/mini-arm64.c:
```diff
@@ -5114,6 +5114,66 @@ emit_store_regset_cfa (MonoCompile *cfg, guint8 *code, guint64 regs, int basereg
+	for (i = 32; i >= 0; --i) {
+		nregs = 1;
+		if (cfa_regset & (1 << i)) {
+			if (pos > 0 && pos % 2 && (cfa_regset & (1 << (i - 1))) && (i - 1 != ARMREG_SP)) {
```
This loop includes `i == 0`, do we need to worry about that case and `i = 1` expression becoming negative?

scott-ferguson-unity replied:
Practically we should never hit that case since that's an arg resister and we shouldn't need to save it, but I've fixed that in https://github.com/Unity-Technologies/mono/pull/1960/commits/9b8e291ee6dde9c51871bf05de6ad195e1d48f01

joncham commented on mono/mini/mini-arm64.c:
```diff
@@ -5114,6 +5114,66 @@ emit_store_regset_cfa (MonoCompile *cfg, guint8 *code, guint64 regs, int basereg
+				}
+			}
+
+			i -= nregs - 1;
```
I'm not following this loop. Perhaps an interactive review makes sense. Otherwise, I would suggest some more comments and some asserts like `i >=0` (assuming that should be true).

scott-ferguson-unity replied:
It's a confusing loop - it's mostly structured to match https://github.com/Unity-Technologies/mono/blob/unity-main/mono/mini/mini-arm64.c#L5089-L5122, otherwise I would have implemented it a bit differently.

joncham commented on mono/mini/exceptions-arm64.c:
```diff
@@ -24,6 +24,12 @@
 #ifndef DISABLE_JIT
 
+#ifdef MONO_ARCH_ENABLE_PTRAUTH
+	static gboolean enable_ptrauth = TRUE;
```
Why is a compile time constant converted to a runtime check? Is that to support cross compile cases?

scott-ferguson-unity replied:
No good reason - I just copied it from https://github.com/Unity-Technologies/mono/blob/unity-main/mono/mini/mini-arm64.c#L278-L292

joncham commented on mono/mini/mini-arm64.c:
```diff
@@ -5134,6 +5194,10 @@ emit_setup_lmf (MonoCompile *cfg, guint8 *code, gint32 lmf_offset, int cfa_offse
 	code = emit_strx (code, ARMREG_LR, ARMREG_FP, lmf_offset + MONO_STRUCT_OFFSET (MonoLMF, pc));
+
+	const int num_save_lmf_instructions = 2;
+	mono_emit_unwind_op_prolog_nop(cfg, code, num_save_lmf_instructions); 
```
Can we assert this constant matches what it corresponds to, i.e. that the lmf operation only uses two instructions? Does this correspond to the operation on line 5360?

scott-ferguson-unity replied:
Changed to use a code size calculation instead of being hard coded in https://github.com/Unity-Technologies/mono/pull/1960/commits/7604e55851a7c96d7e06d1369e732d05a57b0733

joncham commented on mono/mini/exceptions-arm64.c:
```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+static void
+mono_arch_unwind_add_set_fp (guint8* unwind_codes, guint32* unwind_code_size) {
+	mono_arch_unwind_add_assert (1, 0, *unwind_code_size, 0);
+	unwind_codes[(*unwind_code_size)++] = 0b11100001;
```
Nit, a lot of magic numbers in this diff. Not sure how much is to be consistent with other places in Mono, or whether we want to do more.

scott-ferguson-unity replied:
This magic number are from the windows unwind codes - https://learn.microsoft.com/en-us/cpp/build/arm64-exception-handling?view=msvc-170#unwind-codes.  They'll exist only here, not sure that's much benefit in making them defines.

scott-ferguson-unity replied:
I've added comments from the documentation on the format and meaning of these codes in https://github.com/Unity-Technologies/mono/pull/1960/commits/9b8e291ee6dde9c51871bf05de6ad195e1d48f01.  That should help.

joncham commented on mono/mini/exceptions-arm64.c:
```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+
+	mono_arch_unwind_add_assert (2, 0x7FFF, *unwind_code_size, size);
+	int order = reverse ? 1 : 0;
+	size = size / 16; // Size is multiples of 16
```
This will round towards 0. What happens for remainder of size in this case?

scott-ferguson-unity replied:
The stack has to be 16 byte aligned, which is why there's no option to even supply the low 4 bits anywhere.

bholmes replied:
Should we assert if it is not an even multiple?

scott-ferguson-unity replied:
Assert added in https://github.com/Unity-Technologies/mono/commit/9b8e291ee6dde9c51871bf05de6ad195e1d48f01

joncham commented on mono/mini/exceptions-arm64.c:
```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+		mono_arch_unwind_add_pac_sign_lr (unwind_codes, unwind_code_size);
+
+	// Stack probes
+	if (stack_offset + param_area > 0x1000) {
```
This needs kept in sync with `emit_prolog_setup_sp_win64` in `mono/mini/mini-arm64.c`?

scott-ferguson-unity replied:
Yes, basically all of https://github.com/Unity-Technologies/mono/pull/1960/files#diff-90c089fd795716a5ad3b86b5d0fcb22523e846ffd2d286f8189d7e72e081dedcR862-R895 has to be kept in sync.  

I could add more special unwind codes - like `W_CFA_mono_fp_alloc_info_win64_arm64` to track this.  There could be some ordering issues, but it's probably worthwhile

scott-ferguson-unity replied:
And this we already broke by https://github.com/Unity-Technologies/mono/pull/1965

I've fixed this in https://github.com/Unity-Technologies/mono/pull/1960/commits/adb3c97d4d97ce929bdf1f88eeb6eba5d0f17b18.  It's somewhat awkward in that the frame setup gets processed in the unwind code loop.  But this lets us just emit nop's for the stack probes.

@joncham This is a big enough change that I'm requesting a re-review.

joncham reviewed (commented):
Left a few questions, but amazing work!

bholmes commented on mono/mini/exceptions-arm64.c:
```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+static void
+mono_arch_unwind_add_save_reg_x (guint8* unwind_codes, guint32* unwind_code_size, guint32 reg, guint32 offset, gboolean reverse) {
+
+	mono_arch_unwind_add_assert (2, 256, *unwind_code_size, offset);
+
+	guint8 reg_offset = reg - ARMREG_R19;
+	int order = reverse ? 1 : 0;
+
+	unwind_codes[*unwind_code_size + order] = 0b11010000 | reg_offset / 8;
+	unwind_codes[*unwind_code_size + 1 - order] = (reg_offset & 0x8) << 5 | (offset / 8);
+	*(unwind_code_size) += 2;
```
This looks like a copy of `mono_arch_unwind_add_save_reg`. The only diff I notice is the max_offset passed to the assert code.  Is this correct?

scott-ferguson-unity replied:
The binary code, 0b1101000 was wrong - fixed in https://github.com/Unity-Technologies/mono/commit/9b8e291ee6dde9c51871bf05de6ad195e1d48f01

bholmes commented on mono/mini/exceptions-arm64.c:
```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+	int order = reverse ? 1 : 0;
+
+	unwind_codes[*unwind_code_size + order] = 0b11010000 | reg_offset / 8;
+	unwind_codes[*unwind_code_size + 1 - order] = (reg_offset & 0x8) << 5 | (offset / 8);
```
`reg_offset & 0x8` should be `reg_offset & 0x7`?

scott-ferguson-unity replied:
Fixed in https://github.com/Unity-Technologies/mono/commit/9b8e291ee6dde9c51871bf05de6ad195e1d48f01

bholmes commented on mono/mini/exceptions-arm64.c:
```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+mono_arch_unwind_add_alloc_s (guint8* unwind_codes, guint32* unwind_code_size, guint32 size) {
+
+	mono_arch_unwind_add_assert (1, 511, *unwind_code_size, size);
+	unwind_codes[(*unwind_code_size)++] = 0b00000000 | (size / MONO_ARCH_FRAME_ALIGNMENT);
```
Assert that this is aligned?

scott-ferguson-unity replied:
Added in https://github.com/Unity-Technologies/mono/pull/1960/commits/9b8e291ee6dde9c51871bf05de6ad195e1d48f01

scott-ferguson-unity committed:
Fix emit_load_regset_cfa loop and add comments and asserts

bholmes commented on mono/mini/exceptions-arm64.c:
```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+		// We shoulded hit here, but emit a nop if we do
+		g_assert(unwind_op_data->reg >= ARMREG_R19 && unwind_op_data->reg <= ARMREG_R28);
+		mono_arch_unwind_add_nop(unwind_codes, unwind_code_size);
+		return 1;
```
Should we have `mono_arch_unwind_add_*` methods return the size used?  

This becomes `return mono_arch_unwind_add_nop(unwind_codes, unwind_code_size);`.

Just feels safer to me

bholmes replied:
I think I misunderstood the return value. That is the reg count

bholmes commented on mono/mini/exceptions-arm64.c:
```diff
@@ -620,3 +626,598 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+
+			if (reg_count > 0) {
+				if (last_saved_in_order_reg == -1)
+					res.can_use_packed_format = (unwind_op_data->reg == ARMREG_R19);
```
is it possible for `res.can_use_packed_format` to already be false and the result of `(unwind_op_data->reg == ARMREG_R19)` will make it true?  Maybe `if (last_saved_in_order_reg == -1 && res.can_use_packed_format)`

scott-ferguson-unity replied:
Fixed in https://github.com/Unity-Technologies/mono/pull/1960/commits/cec9c68347b719dfbccb1a07e1f859c903bbf4d5

scott-ferguson-unity committed:
Remove unwind code for arch_reg

This was storing a FP offset, not updating FP

scott-ferguson-unity committed:
Fix issues with unwind codes

Fixed some bit manipulation offsets
Fixed the code for save_reg_x
Implement the save unwind code
Fixed bug in saving regs
Added more asserts
Added comments documenting the format of the opcodes

scott-ferguson-unity committed:
Emit uwnind nop count as a calculation, not hard coded.

scott-ferguson-unity committed:
Fixed possible reset of can_use_packed_format

scott-ferguson-unity committed:
Process the ARM64 unwind frame setup in uwnind code loop

By moving this processing into the mono unwind code loop we
can change the stack probe code to emit nop's and still have the
frame windows unwind codes in the correct location.

Also adding support for DW_CFA_def_cfa_offset allows us
to not need to rely on the cfa->stack_offset being the same
as the cfa_offset.  This fixes some use cases with trampolines.

This change also move the setup for the param_area so that it
is done outside of the cfa setup.  This way the native unwind
code now no longer needs to worry about it.

scott-ferguson-unity committed:
Fix emit of epilog codes

We may not have recorded any restore codes
for the registers.  Those are only done if we
stored different registers than we restore.

If that's the case then we need to replay
the DW_CFA_offset codes, but we'll have
to reverse them to match the order that
they are actually emitted.

We can also drop DW_CFA_mono_fp_alloc_info_win64_arm64,
it is not used.

joncham commented on mono/mini/exceptions-arm64.c:
```diff
@@ -620,3 +626,653 @@ mono_arch_do_ip_adjustment (MonoContext *ctx)
+	gboolean has_restore_reg_ops;
+} InitilizedWindowInfoResult;
+
+static InitilizedWindowInfoResult
```
Nit, spelling.

bholmes reviewed (approved):
Good job!  Looked like a fun one.

Did you run some Unity tests locally against this Mono?

scott-ferguson-unity commented:
> Good job! Looked like a fun one.
> 
> Did you run some Unity tests locally against this Mono?

Yes I started Unity and ran the scripting test suite.
